File: app/risk/risk_manager.py
from app.core.logger import logger
from app.services.risk_service import RiskService
from app.portfolio.portfolio_manager import PortfolioManager


class RiskManager:

    @classmethod
    async def validate_order(
        cls,
        symbol: str,
        quantity: float,
        price: float,
        side: str
    ) -> bool:

        settings = await RiskService.get_settings()
        
        # 1. Restricted Symbols Check
        def normalize_sym(s: str) -> str:
            return str(s).strip().upper().replace("/", "").replace("-", "")

        restricted_symbols = [
            normalize_sym(s) 
            for s in settings.get("restricted_symbols", [])
        ]
        if normalize_sym(symbol) in restricted_symbols:
            logger.warning(f"Risk rejected order: {symbol} is restricted")
            return False

        # 2. Max Position Size Check
        order_value = quantity * price
        if order_value > settings.get("max_position_size", 10000.0):
            logger.warning(f"Risk rejected order: size {order_value} exceeds max {settings.get('max_position_size')}")
            return False

        # 3. Max Open Positions Check (Only for BUY orders)
        if side == "BUY":
            open_positions = len(PortfolioManager.positions)
            if open_positions >= settings.get("max_open_positions", 5):
                logger.warning(f"Risk rejected order: max open positions {settings.get('max_open_positions')} reached")
                return False

        # 4. Max Daily Loss / Drawdown Check
        realized_pnl = await PortfolioManager.get_realized_pnl()
        max_daily_loss_limit = -abs(settings.get("max_daily_loss", 5000.0))
        if realized_pnl <= max_daily_loss_limit:
            logger.warning(f"Risk rejected order: max daily loss {max_daily_loss_limit} exceeded")
            return False

        return True

File: backend/src/trading/risk/riskManager.ts
import { logger } from "../../core/logger";
import { RiskService } from "../../services/riskService";
import { PortfolioManager } from "../portfolio/portfolioManager";

export class RiskManager {
  static async validateOrder(
    symbol: string,
    quantity: number,
    price: number,
    side: string
  ): Promise<boolean> {
    const settings = await RiskService.getSettings();

    const normalizeSym = (s: string): string => {
      return String(s).trim().toUpperCase().replace("/", "").replace("-", "");
    };

    // 1. Restricted Symbols Check
    const restrictedSymbols = (settings.restricted_symbols || []).map((s: string) => normalizeSym(s));
    if (restrictedSymbols.includes(normalizeSym(symbol))) {
      logger.warn(`Risk rejected order: ${symbol} is restricted`);
      return false;
    }

    // 2. Max Position Size Check
    const orderValue = quantity * price;
    const maxPositionSize = settings.max_position_size !== undefined ? settings.max_position_size : 10000.0;
    if (orderValue > maxPositionSize) {
      logger.warn(`Risk rejected order: size ${orderValue} exceeds max ${maxPositionSize}`);
      return false;
    }

    // 3. Max Open Positions Check (Only for BUY orders)
    if (side === "BUY") {
      const openPositions = Object.keys(PortfolioManager.positions).length;
      const maxOpenPositions = settings.max_open_positions !== undefined ? settings.max_open_positions : 5;
      if (openPositions >= maxOpenPositions) {
        logger.warn(`Risk rejected order: max open positions ${maxOpenPositions} reached`);
        return false;
      }
    }

    // 4. Max Daily Loss / Drawdown Check
    const realizedPnl = await PortfolioManager.getRealizedPnl();
    const maxDailyLoss = settings.max_daily_loss !== undefined ? settings.max_daily_loss : -5000.0;
    const maxDailyLossLimit = -Math.abs(maxDailyLoss);
    if (realizedPnl <= maxDailyLossLimit) {
      logger.warn(`Risk rejected order: max daily loss ${maxDailyLossLimit} exceeded`);
      return false;
    }

    return true;
  }
}
export default RiskManager;
